### PR TITLE
Refactor fetching retain ports when reconciling SecurityGroup

### PR DESCRIPTION
**Pull request by zarvd**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

- Refactor the function that retrieves the ports for services using the same external IP.
- Reduce a call to fetch PublicIP.
- Add unit test cases.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: internal/testutil/fixture/azure_securitygroup.go [hunk 1]
```diff
@@ -24,16 +24,20 @@ import (
...
 	"sigs.k8s.io/cloud-provider-azure/pkg/provider/loadbalancer/fnutil"
 	"sigs.k8s.io/cloud-provider-azure/pkg/provider/loadbalancer/iputil"
 	"sigs.k8s.io/cloud-provider-azure/pkg/provider/loadbalancer/securitygroup"
-	"sigs.k8s.io/cloud-provider-azure/pkg/provider/loadbalancer/testutil"
 )
 
 // NoiseSecurityRules returns N non cloud-provider-specific security rules.
 // It's not random, but it's good enough for testing.
-func (f *AzureFixture) NoiseSecurityRules(nRules int) []network.SecurityRule {
+func (f *AzureFixture) NoiseSecurityRules() []network.SecurityRule {
```
Comment (nilo19) [inline internal/testutil/fixture/azure_securitygroup.go:36]: Do we want to change the func name or the comments above?

File: internal/testutil/fixture/azure_securitygroup.go [hunk 2]
```diff
@@ -24,16 +24,20 @@ import (
...
+	return f.NNoiseSecurityRules(3)
+}
+
+func (f *AzureFixture) NNoiseSecurityRules(nRules int) []network.SecurityRule {
```
Comment (nilo19) [inline internal/testutil/fixture/azure_securitygroup.go:40]: Better to have comments for each exported function.

File: pkg/provider/azure_loadbalancer.go [hunk 3]
```diff
@@ -2919,17 +2814,12 @@ func (az *Cloud) reconcileSecurityGroup(
 		WithValues("delete-lb", !wantLb)
 	logger.V(2).Info("Starting")
 
+	ctx := klog.NewContext(context.Background(), logger)
```
Comment (nilo19) [inline pkg/provider/azure_loadbalancer.go:2817]: How about using context from outside caller?

**Comment by nilo19** [comment]
Do you consider internal/testutil -> testutil/?

File: internal/testutil/fixture/azure_securitygroup.go (hunk 1 above)
Comment (zarvd) [reply internal/testutil/fixture/azure_securitygroup.go:36]: updated, thanks

File: pkg/provider/azure_loadbalancer.go (hunk 3 above)
Comment (zarvd) [reply pkg/provider/azure_loadbalancer.go:2817]: I am thinking so, will do it in another logging refactor PR.

**Comment by zarvd** [comment]
> Do you consider internal/testutil -> testutil/?

I am thinking not to export this pkg, and maybe some more

**Comment by nilo19** [comment]
> > Do you consider internal/testutil -> testutil/?
> 
> I am thinking not to export this pkg, and maybe some more

Then I think we should include this change within the pr.

**Pull request merged**